reactor-netty-http/src/main/java/reactor/netty/http/Http2ConnectionLiveness.java

```diff
@@ -0,0 +1,225 @@
+ * explicit synchronization. A new instance is created per channel.</p>
+ *
+ * <h3>Configuration Guidelines</h3>
+ * <ul>
+ *   <li><strong>pingAckTimeout:</strong> Should account for expected network latency and load.
+ *       Values that are too short may cause false positives due to temporary delays.</li>
+ *   <li><strong>pingAckDropThreshold:</strong> Balance between quick failure detection and tolerance
+ *       for transient issues. Higher values increase detection time but reduce false positives.</li>
+ * </ul>
+ *
+ * @author raccoonback
+ * @author Violeta Georgieva
+ * @since 1.2.12
+ */
+public final class Http2ConnectionLiveness implements HttpConnectionLiveness {
+
+	private final Http2FrameWriter http2FrameWriter;
+	private final int pingAckDropThreshold;
+	private final long pingAckTimeoutNanos;
+
+	private boolean isPingAckPending;
+	private long lastSentPingData;
+	private ScheduledFuture<?> pingScheduler;
+	private int pingAttempts;
+
+	/**
+	 * Constructs a new {@code Http2ConnectionLiveness} instance.
+	 *
+	 * @param http2FrameCodec the HTTP/2 frame codec
+	 * @param pingAckDropThreshold the maximum number of PING frame transmission attempts before closing the connection
+	 * @param pingAckTimeoutNanos the timeout in nanoseconds for receiving a PING ACK response
+	 */
+	public Http2ConnectionLiveness(Http2FrameCodec http2FrameCodec, int pingAckDropThreshold, long pingAckTimeoutNanos) {
+		this.http2FrameWriter = http2FrameCodec.encoder().frameWriter();
+		this.pingAckDropThreshold = pingAckDropThreshold;
+		this.pingAckTimeoutNanos = pingAckTimeoutNanos;
+	}
+
+	/**
+	 * Cancels the scheduled ping task.
+	 */
+	@Override
+	public void cancel() {
+		isPingAckPending = false;
+		pingAttempts = 0;
+		if (pingScheduler != null) {
+			pingScheduler.cancel(false);
+			pingScheduler = null;
+		}
+	}
+
+	/**
+	 * Checks the liveness of the connection and schedules a ping if necessary.
+	 *
+	 * @param ctx the {@link ChannelHandlerContext} of the connection
+	 */
+	@Override
+	@SuppressWarnings("FutureReturnValueIgnored")
+	public void check(ChannelHandlerContext ctx) {
+		if (!isPingAckPending) {
+			if (log.isDebugEnabled()) {
+				log.debug(format(ctx.channel(), "Connection was idle. Starting probing with PING frame: timeout={} ns."),
+						pingAckTimeoutNanos);
+			}
+
+			PingTimeoutTask pingTimeoutTask = new PingTimeoutTask(ctx);
+			pingTimeoutTask.writePing();
+		}
+	}
+
+	/**
+	 * Receives a message from the peer and processes it if it is a ping frame.
+	 *
+	 * @param msg the message received from the peer
+	 */
+	@Override
+	public void receive(Object msg) {
+		if (isPingAckPending) {
+			if (msg instanceof Http2PingFrame) {
+				Http2PingFrame frame = (Http2PingFrame) msg;
+				if (frame.ack() && frame.content() == lastSentPingData) {
+					cancel();
+				}
+			}
+			else if (msg instanceof Http2HeadersFrame || msg instanceof Http2DataFrame) {
+				cancel();
+			}
```

Comment:
Oh, so it also handles the case when a header frame is received. 👍

Reply:
yeah I think headers/data defines a request/response